### PR TITLE
Implemented failover detector

### PR DESCRIPTION
**Pull request by balavor**
## Changes
- Implemented Failover detector;
- Removed content buffering restriction for live video;
- Added more tests in vrm group processing controller

@VerizonAdPlatforms/video-partner-sdk-developers: Please review.

### CODE REVIEW

File: sources/metrics/detectors/failover/FailoverDetector.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//  Copyright 2018, Oath Inc.
+//  Licensed under the terms of the MIT License. See LICENSE.md file in project root for terms.
+import Foundation
+
+extension Detectors {
+    final class Failover {
+        
+        var adSessionID = UUID()
```
Comment (VladyslavAnokhin) [inline sources/metrics/detectors/failover/FailoverDetector.swift]: private?

File: sources/metrics/detectors/failover/FailoverDetectorTests.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        XCTAssertEqual(result, false)
+        
+        result = detector.process(isFailover: true, adSessionID: adSessionID)
+        XCTAssertEqual(result, true)
+        
+        result = detector.process(isFailover: true, adSessionID: adSessionID)
+        XCTAssertEqual(result, false)
+        
+        result = detector.process(isFailover: true, adSessionID: UUID())
+        XCTAssertEqual(result, true)
```
Comment (VladyslavAnokhin) [inline sources/metrics/detectors/failover/FailoverDetectorTests.swift]: You can replace all this assertions with bool assertions

File: sources/metrics/detectors/failover/FailoverDetector.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        
+        func process(vrmResponse: VRMResponse?,
+                     adSessionID: UUID?) -> AdResult {
+            guard let vrmResponse = vrmResponse,
+                let adSessionID = adSessionID,
+                transactionIds.contains(adSessionID) == false else {
+                currentResult = .other
+                return .other
+            }
+            if vrmResponse.groups.isEmpty && currentResult == .other {
```
Comment (AndriiMoskvin) [inline sources/metrics/detectors/failover/FailoverDetector.swift]: Why not place it in `switch`?

**Review by AndriiMoskvin — DISMISSED** [review]

File: sources/metrics/detectors/failover/FailoverDetector.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
+//  Copyright 2018, Oath Inc.
+//  Licensed under the terms of the MIT License. See LICENSE.md file in project root for terms.
+import Foundation
+import PlayerCore
+
+extension Detectors {
+    final class Failover {
+        
+        enum AdResult {
```
Comment (AndriiMoskvin) [inline sources/metrics/detectors/failover/FailoverDetector.swift]: Can we extend this `enum` with more state? Use more type driven approach.

File: sources/metrics/detectors/failover/FailoverDetector.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
+//  Copyright 2019, Oath Inc.
+//  Licensed under the terms of the MIT License. See LICENSE.md file in project root for terms.
+import Foundation
+import PlayerCore
+
+extension Detectors {
+    final class Failover {
+        
+        private enum AdResult {
+            case other
```
Comment (VladyslavAnokhin) [inline sources/metrics/detectors/failover/FailoverDetector.swift]: `other` -> `none` ?

File: sources/metrics/detectors/failover/FailoverDetector.swift [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                adSessionIDs.insert(adSessionID)
+                return false
+            }
+            guard currentGroup == nil,
+                groupQueue.isEmpty else {
+                    currentResult = .processing
+                    return false
+            }
+            currentResult = .failover
+            adSessionIDs.insert(adSessionID)
```
Comment (AndriiMoskvin) [inline sources/metrics/detectors/failover/FailoverDetector.swift]: `currentResult = smth and insert of ID` can be moved to separate function and reused.

**Review by AndriiMoskvin — DISMISSED** [review]

File: sources/metrics/detectors/failover/FailoverDetectorTests.swift [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                                      adSessionID: adSessionID)
+        XCTAssertFalse(result)
+    }
+    
+    func testFailoverCase() {
+        var result = detector.process(vrmResponse: response,
+                                      currentGroup: group,
+                                      groupQueue: [group],
+                                      adSessionID: adSessionID)
+        result = detector.process(vrmResponse: response,Г
```
Comment (VladyslavAnokhin) [inline sources/metrics/detectors/failover/FailoverDetectorTests.swift]: `Г` :D

File: sources/metrics/detectors/failover/FailoverDetector.swift [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            guard vrmResponse.groups.isEmpty == false,
+               currentResult != .noAds else {
+                finishProcessing(with: .noAds)
+                return false
+            }
+            guard currentGroup == nil,
+                groupQueue.isEmpty else {
+                    return false
+            }
+            finishProcessing(with: .noAds)
```
Comment (VladyslavAnokhin) [inline sources/metrics/detectors/failover/FailoverDetector.swift]: `.noAds` -> `.failover`

**Review by VladyslavAnokhin — DISMISSED** [review]
See my comment

File: sources/metrics/detectors/failover/FailoverDetector.swift [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                     currentGroup: VRMCore.Group?,
+                     groupQueue: [VRMCore.Group],
+                     adSessionID: UUID?) -> Bool {
+            guard let vrmResponse = vrmResponse,
+                let adSessionID = adSessionID,
+                adSessionIDs.contains(adSessionID) == false else {
+                    currentResult = .none
+                    return false
+            }
+            func finishProcessing(with result: AdResult) {
```
Comment (AndriiMoskvin) [inline sources/metrics/detectors/failover/FailoverDetector.swift]: Incorporate return `true/false` in this function?

**Review by AndriiMoskvin — DISMISSED** [review]

**Review by VladyslavAnokhin — APPROVED** [review]

**Pull request merged**